File: src/main/java/gargoyle/ct/config/CTConfigs.java
package gargoyle.ct.config;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.InvalidObjectException;
import java.io.Serializable;
import java.text.MessageFormat;
import java.util.*;

public class CTConfigs implements Serializable {
    private static final String MSG_NOT_VALID_CONVERT_0 = "not valid convert: {0}";
    private static final long serialVersionUID = 2024075953874239351L;
    private final Map<String, CTConfig> configs = new LinkedHashMap<>();

    public CTConfigs(@NotNull CTConfig... configs) {
        setConfigs(configs);
    }

    private void setConfigs(CTConfig[] configs) {
        synchronized (this.configs) {
            this.configs.clear();
            for (CTConfig config : configs) {
                addConfig(config);
            }
        }
    }

    public void addConfig(@NotNull CTConfig config) {
        synchronized (configs) {
            String name = config.getName();
            if (!configs.containsKey(name)) {
                configs.put(name, config);
            }
        }
    }

    @SuppressWarnings("TypeMayBeWeakened")
    public CTConfigs(@NotNull List<CTConfig> configs) {
        setConfigs(configs);
    }

    public CTConfig getConfig(String name) {
        return configs.get(name);
    }

    @NotNull
    public List<CTConfig> getConfigs() {
        return Collections.unmodifiableList(new LinkedList<>(configs.values()));
    }

    private void setConfigs(Iterable<CTConfig> configs) {
        synchronized (this.configs) {
            this.configs.clear();
            for (CTConfig config : configs) {
                addConfig(config);
            }
        }
    }

    public boolean hasConfig(@NotNull CTConfig config) {
        return configs.containsKey(config.getName());
    }

    @Override
    public int hashCode() {
        int prime = 31;
        int result = 1;
        result = prime * result + configs.hashCode();
        return result;
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        CTConfigs other = (CTConfigs) obj;
        return Objects.equals(configs, other.configs);
    }

    @NotNull
    @Override
    public String toString() {
        return MessageFormat.format("CTConfigs [configs={0}]", configs);
    }

    public void validate() throws InvalidObjectException {
        for (CTConfig config : configs.values()) {
            if (config == null || config.isNotValid()) {
                throw new InvalidObjectException(MessageFormat.format(MSG_NOT_VALID_CONVERT_0, config));
            }
        }
    }
}

File: app/src/main/kotlin/gargoyle/ct/config/CTConfigs.kt
package gargoyle.ct.config

import gargoyle.ct.config.convert.impl.CTConfigsConverter
import kotlinx.serialization.Serializable
import java.text.MessageFormat
import java.util.*

@Serializable(with = CTConfigsConverter::class)
open class CTConfigs {
    private val configs: MutableMap<String, CTConfig> = LinkedHashMap()

    constructor(vararg configs: CTConfig) {
        setConfigs(arrayOf(*configs))
    }

    private fun setConfigs(configs: Array<CTConfig>) {
        synchronized(this.configs) {
            this.configs.clear()
            for (config in configs) {
                addConfig(config)
            }
        }
    }

    fun addConfig(config: CTConfig) {
        synchronized(configs) {
            if ((config.name) !in configs) configs[config.name] = config
        }
    }

    constructor(configs: List<CTConfig>) {
        setConfigs(configs)
    }

    fun getConfig(name: String): CTConfig? = configs[name]

    fun getConfigs(): List<CTConfig> = LinkedList(configs.values)

    private fun setConfigs(configs: Iterable<CTConfig>) {
        synchronized(this.configs) {
            this.configs.clear()
            configs.forEach { addConfig(it) }
        }
    }

    fun hasConfig(config: CTConfig): Boolean = configs.containsKey(config.name)

    override fun hashCode(): Int = 31 * 1 + configs.hashCode()

    override fun equals(other: Any?): Boolean =
        when {
            this === other -> true
            other == null -> false
            javaClass != other.javaClass -> false
            else -> configs == (other as CTConfigs).configs
        }

    override fun toString(): String = MessageFormat.format("CTConfigs [configs={0}]", configs)

    fun validate() {
        configs.values.first { !it.isValid }
    }

    companion object {
        private const val MSG_NOT_VALID_CONVERT_0 = "not valid convert: {0}"
    }
}
